internal/pkg/push/push.go

```diff
@@ -20,25 +20,135 @@ package push
 
 import (
 	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/protobom/protobom/pkg/sbom"
 
 	"github.com/bomctl/bomctl/internal/pkg/client"
+	"github.com/bomctl/bomctl/internal/pkg/db"
+	"github.com/bomctl/bomctl/internal/pkg/fetch"
 	"github.com/bomctl/bomctl/internal/pkg/options"
 )
 
 func Push(sbomID, destPath string, opts *options.PushOptions) error {
 	opts.Logger.Info("Pushing Document", "sbomID", sbomID)
 
-	pusher, err := client.New(destPath)
+	// create appropriate push client based on user provided destination
+	pClient, err := client.New(destPath)
 	if err != nil {
 		return fmt.Errorf("creating push client: %w", err)
 	}
 
-	opts.Logger.Info(fmt.Sprintf("Pushing to %s URL", pusher.Name()), "url", destPath)
+	opts.Logger.Info(fmt.Sprintf("Pushing to %s URL", pClient.Name()), "url", destPath)
 
-	err = pusher.Push(sbomID, destPath, opts)
+	// push sbomid to destpath via client
+	err = pClient.Push(sbomID, destPath, opts)
 	if err != nil {
 		return fmt.Errorf("failed to push to %s: %w", destPath, err)
 	}
 
+	// If user wants to recurse the sbom tree and push all, do so
+	if opts.UseTree {
+		err := getExtRefDocs(sbomID, destPath, opts)
+		if err != nil {
+			return fmt.Errorf("failed to fetch external ref boms for %s: %w", sbomID, err)
+		}
+	}
+
+	return nil
+}
+
+func getExtRefDocs(sbomID, destPath string, opts *options.PushOptions) error {
+	opts.Logger.Info("Fetching External Ref Boms for Document", "sbomID", sbomID)
+
+	backend, err := db.BackendFromContext(opts.Context())
+	if err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	extRefs, err := backend.GetExternalReferencesByDocumentID(sbomID, "BOM")
+	if err != nil {
+		return fmt.Errorf("error getting external references: %w", err)
+	}
+
+	// for each external ref fetch and push external ref bom document
+	for _, ref := range extRefs {
+		err := pushExtRefDoc(ref, backend, destPath, opts)
+		if err != nil {
+			return fmt.Errorf("error pulling external ref file: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func checkForExistingDoc(be *db.Backend, doc *sbom.Document) (id, name string, err error) {
```

Comment:
So I didnt feel like get was appropriate, since if the doc fetched doesn't exist in the db, then it returns info for the doc that was passed in. maybe just `getDocInfo`?

Reply:
I think the alternative sounds solid.